diff --git a/src/smpi/internals/smpi_global.cpp b/src/smpi/internals/smpi_global.cpp
index fc23cde89..f767ffc55 100644
--- a/src/smpi/internals/smpi_global.cpp
+++ b/src/smpi/internals/smpi_global.cpp
@@ -611,6 +611,9 @@ void SMPI_init(){
       process_data.erase(it);
     }
   });
+  simgrid::s4u::Host::on_creation.connect(
+    [](simgrid::s4u::Host& host) { host.extension_set(new simgrid::smpi::Host(&host));}
+  );
 
   smpi_init_options();
   smpi_global_init();
